core/src/trezor/wire/thp/received_message_handler.py

```diff
@@ -0,0 +1,498 @@
+
+
+def _send_ack(ctx: Channel, ack_bit: int) -> Awaitable[None]:
+    ctrl_byte = control_byte.add_ack_bit_to_ctrl_byte(ACK_MESSAGE, ack_bit)
+    header = PacketHeader(ctrl_byte, ctx.get_channel_id_int(), CHECKSUM_LENGTH)
+    if __debug__:
+        log.debug(
+            __name__,
+            "Writing ACK message to a channel with cid: %s, ack_bit: %d",
+            hexlify_if_bytes(ctx.channel_id),
+            ack_bit,
+            iface=ctx.iface,
+        )
+    return write_payload_to_wire_and_add_checksum(ctx.iface, header, b"")
+
+
+def _check_checksum(message_length: int, message_buffer: utils.BufferType) -> None:
+    if __debug__:
+        log.debug(__name__, "check_checksum")
+    if not checksum.is_valid(
+        checksum=message_buffer[message_length - CHECKSUM_LENGTH : message_length],
+        data=memoryview(message_buffer)[: message_length - CHECKSUM_LENGTH],
+    ):
+        if __debug__:
+            log.debug(__name__, "Invalid checksum, ignoring message.")
+        raise ThpError("Invalid checksum, ignoring message.")
+
+
+async def handle_ack(ctx: Channel, ack_bit: int) -> None:
+    if not ABP.is_ack_valid(ctx.channel_cache, ack_bit):
+        return
+    # ACK is expected and it has correct sync bit
+    if __debug__:
+        log.debug(
+            __name__,
+            "Received ACK message with correct ack bit",
+            iface=ctx.iface,
+        )
+    if ctx.transmission_loop is not None:
+        ctx.transmission_loop.stop_immediately()
+        if __debug__:
+            log.debug(__name__, "Stopped transmission loop", iface=ctx.iface)
+    elif __debug__:
+        log.debug(__name__, "Transmission loop was not stopped!", iface=ctx.iface)
+
+    ABP.set_sending_allowed(ctx.channel_cache, True)
+
+    if ctx.write_task_spawn is not None:
+        if __debug__:
+            log.debug(
+                __name__,
+                'Control to "write_encrypted_payload_loop" task',
+                iface=ctx.iface,
+            )
+        await ctx.write_task_spawn
+        # Note that no the write_task_spawn could result in loop.clear(),
+        # which will result in termination of this function - any code after
+        # this await might not be executed
+
+
+def _handle_message_to_app_or_channel(
+    ctx: Channel,
+    payload_length: int,
+    message_length: int,
+    ctrl_byte: int,
+) -> Awaitable[None]:
+    state = ctx.get_channel_state()
+
+    if state is ChannelState.ENCRYPTED_TRANSPORT:
+        return _handle_state_ENCRYPTED_TRANSPORT(ctx, message_length)
+
+    if state is ChannelState.TH1:
+        return _handle_state_TH1(ctx, payload_length, message_length, ctrl_byte)
+
+    if state is ChannelState.TH2:
+        return _handle_state_TH2(ctx, message_length, ctrl_byte)
+
+    if _is_channel_state_pairing(state):
+        return _handle_pairing(ctx, message_length)
+
+    raise ThpError("Unimplemented channel state")
+
+
+async def _handle_state_TH1(
```

Comment:
nit: this function doesn't `await`, so it probably doesn't have to be `async`
(here and below)